File: algorithms/python/integertoRoman/integertoRoman.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/integer-to-roman/
# Author: Miao Zhang
# Date:   2021-01-05

class Solution:
    def intToRoman(self, num: int) -> str:
        dicts = {
            1:    'I',
            4:    'IV',
            5:    'V',
            9:    'IX',
            10:   'X',
            40:   'XL',
            50:   'L',
            90:   'XC',
            100:  'C',
            400:  'CD',
            500:  'D',
            900:  'CM',
            1000: 'M'
        }
        
        res = ''
        for key in sorted(dicts.keys())[::-1]:
            tmp = num // key
            if tmp == 0:
                continue
            res += tmp * dicts[key]
            num -= tmp * key
            if num == 0:
                break
        return res

File: algorithms/cpp/integertoRoman/integertoRoman.cpp
// Source: https://leetcode.com/problems/integer-to-roman/
// Author: Miao Zhang
// Date:   2021-01-05

class Solution {
public:
    string intToRoman(int num) {
        vector<string> symbol = {"M","CM","D","CD","C","XC","L","XL","X","IX","V","IV","I"};
        vector<int> value = {1000,900,500,400, 100, 90,  50, 40,  10, 9,   5,  4,   1};
        string res;
        
        for (int i = 0; num != 0; i++) {
            while (num >= value[i]) {
                num -= value[i];
                res += symbol[i];
            }
        }
        
        return res;
    }
};
